Mixed Reality Toolkit Standard Shader New Features

Cameron-Micka opened the pull request:
Overview
---
This is an update to the MRTK/Standard shader and systems. The primary change is the addition of per pixel metallic, smoothness, emissive, and occlusion control via channel maps. For example:

![channelmap](https://user-images.githubusercontent.com/13305729/43346530-43e88a18-91a6-11e8-8a52-e1c1a60fd8c3.gif)

Built in configurable stencil test support to achieve a wide array of effects. Such as portals:

![stenciltest](https://user-images.githubusercontent.com/13305729/43346556-621f9bc0-91a6-11e8-90fa-0d4015003248.gif)

Instanced color support to give thousands of GPU instanced meshes unique material properties (example scene not in this PR):

![instances](https://user-images.githubusercontent.com/13305729/43346720-243d838e-91a7-11e8-82a3-cbb082052039.gif)


Finally, a scene to compare and test the MRTK/Standard shader against the Unity/Standard shader:

![comparison](https://user-images.githubusercontent.com/13305729/43346748-4640f754-91a7-11e8-927e-60ed4f98e010.gif)

Single pass instanced stereo rendering has also been enabled by default and a few minor MRTK/Standard shader bugs have been fixed.

Changes
---
- Fixes: #2298 
- Fixes: #1769

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -12,24 +13,37 @@ namespace Microsoft.MixedReality.Toolkit.Internal.Utilities
     public class ClipPlane : MonoBehaviour
     {
+        /// <summary>
+        /// The renderer(s) that should be affected by the clip plane.
```
Don't need a summary tag on the private field. Just a public property if one exists.

Cameron-Micka replied:
Opps, I'll fix this.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -62,95 +76,114 @@ private void OnDrawGizmosSelected()
-            materials = new Material[renderers.Length];
 
-            for (int i = 0; i < renderers.Length; ++i)
+            foreach (Renderer renderer in Renderers)
```
for loop change?

davidkline-ms replied:
we have standardized on for vs foreach. consistency is good :)

Cameron-Micka replied:
I'm making the requested changes now, I'm confused on what to change here?

StephenHodgson replied:
let's revert it back to the for loop for perf

Cameron-Micka replied:
Ah okay, I'll make that change to conform to the standard. Is the perf worry about allocations? If so, I believe Unity fixed that in 5.6: https://forum.unity.com/threads/upgraded-c-compiler-on-5-3-5p8.417363/ "this generally means that doing 'foreach' over an Array, List<T>, Dictionary<T>, or HashSet<T> should not allocate."

StephenHodgson replied:
Yes mainly for allocations, but even on platforms like HoloLens, the perf difference is pretty noticeable in certain hot code areas.

Cameron-Micka replied:
Sounds good!

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -62,95 +76,114 @@ private void OnDrawGizmosSelected()
         private void UpdatePlanePosition()
         {
-            if (renderers == null)
+            if (Renderers == null)
```
let's continue to reference the local field.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -62,95 +76,114 @@ private void OnDrawGizmosSelected()
+            }
+            else
+            {
+                return renderer.material;
```
this creates a new copy of that material. Are we disposing that material correctly to avoid the memory leak?

Cameron-Micka replied:
Ah, no we aren't. I'll loop though modifiedMaterials OnDestroy and clean them up there.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/Demos/StandardShader/Scripts/Editor/MaterialMatrixEditor.cs:
```diff
@@ -0,0 +1,30 @@
+    /// Editor to build a matrix of spheres demonstrating a spectrum of material properties.
+    /// </summary>
+    [CustomEditor(typeof(MaterialMatrix))]
+    public class MaterialMatrixEditor : Editor
```
This may not work as expected because of the custom assembly definitions we're using.

StephenHodgson replied:
Nvm, I see you added one. Great!

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/Demos/StandardShader/Scripts/Editor/MixedRealityToolkit.Examples.Demos.StandardShader.Editor.asmdef:
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "MixedRealityToolkit.Examples.Demos.StandardShader.Editor",
```
We've been sticking to the `Inspectors` terminology throughout the rest of the MRTK. Do you mind doing the same?

Cameron-Micka replied:
Not at all, I'll change that!

SimonDarksideJ commented:
Are we any closer to a LRP supported shader with this update @Cameron-Micka ? or is it looking like we'll need an LRP variant of our standard shader.

P.S. Awesome work, you are a shader god :D

davidkline-ms commented:
> P.S. Awesome work, you are a shader god :D

Seconded!

StephenHodgson commented:
https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2521#issuecomment-408998420

SimonDarksideJ commented:
Agreed, we only need demo features to show MRTK features. 
if it's for comparison, have it in a private repo to demonstrate it and use screenshots in the demo documentation.

Cameron-Micka commented:
Thank you all for the kind words! :) I'm making the requested changes now.

davidkline-ms commented on Assets/MixedRealityToolkit-Examples/Demos/StandardShader/Scripts/MaterialMatrix.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            List<Transform> children = transform.Cast<Transform>().ToList();
+
+            foreach (Transform child in children)
```
foreach -> for

Cameron-Micka replied:
Thank you for the offer, I wish I would have seen this earlier. But, I think I've just made all the requested changes.

davidkline-ms commented:
> What's the difference between Material Gallery and Standard Materials? Can we get rid of one of these scenes?

The purpose for the Standard Materials scene is to give AR developers (ex: HoloLens) and opportunity to see what the materials really look like on their device. During design/development getting on the device is paramount to selecting the best materials for the experience. 

For VR devices, having a simple screen shot in the docs or using the editor to preview materials is generally sufficient.

StephenHodgson commented:
> The purpose for the Standard Materials scene is to give AR developers (ex: HoloLens) and opportunity to see what the materials really look like on their device.

I'm totally for this, but still I feel like both scenes do pretty much the same thing.

davidkline-ms commented:
> I'm totally for this, but still I feel like both scenes do pretty much the same thing.

The main difference is that the Std Materials scene is about choosing good colors, the Material Gallery is about showing what the shader is capable of.

Two different things to me.

Cameron-Micka commented:
I personally like having the Standard Materials scene for picking colors, if there is some concern about too many scenes we could merge the Material Gallery and Standard Materials scenes?

Cameron-Micka commented:
@SimonDarksideJ Thank you! I plan to look at LRP next, since these changes started getting larger and larger I wanted to focus on just this first.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -168,7 +178,14 @@ private Material GetMaterial(Renderer renderer)
             else
             {
-                return renderer.material;
+                Material material = renderer.material;
```
Any time you call `renderer.material` you generate another material. You should just be able to loop the list and destroy them as is.

Cameron-Micka replied:
Unity only generates a new material the first time you call `renderer.material` then it is smart enough to realize an instance has already been created and returns that for later calls.

StephenHodgson replied:
> then it is smart enough to realize an instance has already been created and returns that for later calls

You'd think that, but it's actually the worst memory leak in Unity.

https://docs.unity3d.com/ScriptReference/Renderer-material.html

>Note:
This function automatically instantiates the materials and makes them unique to this renderer. **It is your responsibility to destroy the materials when the game object is being destroyed**. Resources.UnloadUnusedAssets also destroys the materials but it is usually only called when loading a new level.

[Use `renderer.GetMaterials()` instead](Renderer.GetMaterials)

keveleigh replied:
From your first link:

>If the material is used by any other renderers, this will clone the shared material and start using it from now on.

StephenHodgson replied:
Indeed, the cloned instance still needs to be cleaned up.

keveleigh replied:
Is this not doing that?

https://github.com/Cameron-Micka/MixedRealityToolkit-Unity/blob/8499045c2f442a2d840c6aa5f04500e8a75615ec/Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs#L107

Cameron-Micka replied:
I just tried this in an empty project:

    public Renderer testRenderer;

    private void Update()
    {
        if (Input.anyKeyDown)
        {
            Material material = testRenderer.material;
            material.color = Color.red;
        }
    }

And, my material count went up by one (Window->Profiler->Memory->Materials) the first time any key was hit, then maintained constant (material count and memory wise) with sequential key strokes. @StephenHodgson are you see a different behavior?

Cameron-Micka replied:
It's better to be paranoid than leaking!

StephenHodgson replied:
> are you see a different behavior?

Nope I'm seeing the same.

keveleigh commented on Assets/MixedRealityToolkit-Examples/Demos/StandardShader/Scripts/Editor/MixedRealityToolkit.Examples.Demos.StandardShader.Inspectors.asmdef:
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "MixedRealityToolkit.Examples.Demos.StandardShader.Inspectors",
```
Please update this to match our other asmdef names (they start with `Microsoft.MixedReality.Toolkit` instead of `MixedRealityToolkit`). Probably `Microsoft.MixedReality.Toolkit.Examples.Demos.StandardShader.Inspectors` (though that's a little long haha).

Cameron-Micka replied:
Haha, sounds perfect to me. I'll fix that!

keveleigh commented on Assets/MixedRealityToolkit-Examples/Demos/StandardShader/Scenes/MaterialGallery.unity:
```diff
@@ -115,7 +115,7 @@ NavMeshSettings:
   m_NavMeshData: {fileID: 0}
 --- !u!4 &7969012 stripped
 Transform:
-  m_PrefabParentObject: {fileID: 4085006294760784, guid: 6e6521182b532d245b0d80293b332399,
```
Not sure what the issue is here, but I'm missing many prefab references in basically all the scenes. It doesn't like any guids have changed though.

![image](https://user-images.githubusercontent.com/3580640/43491665-47f9f600-94da-11e8-9ff2-3e46c6691e3d.png)

Cameron-Micka replied:
I just restarted Unity and I'm seeing this too, which is very concerning... let me se what is going on.

Cameron-Micka replied:
I was just about to ask if anyone has seen this, the guids look the same. And, I haven't touched those prefabs... has anyone seen something like this before? I'm going to keep digging.

keveleigh replied:
Yeah, I'm not seeing any messed up metas on my end either. I tried regenerating the Library folder with no success.

Cameron-Micka replied:
I need to step out for today, but I did notice it appears Unity changed the serialized names of 3 prefab tracking variables (looks like an innocent name change) in the scene assets. Reverting the scene restores the prefab linkage, but I would like to see if I can find a proper fix.

Cameron-Micka replied:
Found a real fix, I want to test one thing then I'll push the changes. This was the problem for those interested: 

I needed to replace 4 text strings in the scene: 
"serializedVersion: 6" => "serializedVersion: 5" 
"m_CorrespondingSourceObject" => "m_PrefabParentObject" 
"m_IsPrefabAsset" => "m_IsPrefabParent" 
"m_SourcePrefab" => "m_ParentPrefab"


Source: https://answers.unity.com/questions/1520106/what-does-assertion-failed-on-expression-prefabisp.html

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs:
```diff
@@ -25,6 +23,7 @@ public class ClipPlane : MonoBehaviour
         private Dictionary<Material, bool> modifiedMaterials = new Dictionary<Material, bool>();
+        private List<Material> allocatedMaterials = new List<Material>();
 
         public Renderer[] Renderers => renderers;
```
Do you mind if we co-locate the property and field together?

We also might want to put the tooltip string into a summary tag for the renderers as well.

Cameron-Micka replied:
I can do that!

Cameron-Micka commented:
Okay, I believe I've addressed all of the reviews now! If someone could pull down the latest changes and make sure the prefab linkage is fixed for them as well, that would be much appreciated.

StephenHodgson commented:
>  If someone could pull down the latest changes and make sure the prefab linkage is fixed for them as well, that would be much appreciated.

Will do!

StephenHodgson commented:
@Cameron-Micka unfortunately I'm still getting the missing  prefab references.

Cameron-Micka commented:
😢 Thanks for looking, let me do a git clean and see if it comes back for me.

Cameron-Micka commented:
@StephenHodgson I just did a clean and the prefab references are still good. Out of curiosity which editor version are you using?

StephenHodgson commented:
2018.1.9

Cameron-Micka commented:
Yay! Thank you everyone for reviewing this.

StephenHodgson commented on ProjectSettings/ProjectSettings.asset:
```diff
@@ -46,7 +46,7 @@ PlayerSettings:
   defaultScreenWidthWeb: 960
   defaultScreenHeightWeb: 600
-  m_StereoRenderingPath: 0
+  m_StereoRenderingPath: 2
```
@Cameron-Micka I hadn't noticed we changed this. Any chance we could put it back to multipass?

The Single Pass Instanced is still in preview

Cameron-Micka replied:
I can certainly switch it back, or move to normal Single Pass? My team has been using Single Pass Instanced for all of our projects (even enterprise applications delivered to clients) and it has been super stable and a big perf win.

I think the “preview” portion is a handful of Unity extensions haven’t been updated to utilize it. Such as TextMesh Pro. But, it’s pretty easy to fix those instances. (No pun intended. 😉)

StephenHodgson replied:
I'm getting a slight stutter every few seconds with single pass enabled.

Cameron-Micka replied:
Ah interesting, are you seeing it in a particular scene or all scenes? I suppose we have only been using Unity 2017 LTS so haven’t done much testing in 2018.

Cameron-Micka replied:
It looks like #2521 might change this setting? But, please let me know if there is anything you need me to do.

StephenHodgson replied:
Yup, I had requested @keveleigh to change it back. Thanks 👍

isurfraz commented:
Thanks for this, this looks fantastic!

isurfraz commented:
@Cameron-Micka is this in current release 2017.4.1, and if not is there a way we could get this backported to the current MRTK, or as a unitypackage hot-release?

StephenHodgson commented:
We should definitely look into back porting it to the HTK

Cameron-Micka commented:
I would love to back port these features to the HTK. I'm short on time currently, but I can put this on my "todo" list or if someone else want's to port it I'm happy to review the PR. :)

isurfraz commented:
@Cameron-Micka @StephenHodgson I have brought this into our project that uses an earlier version of MRTK (holotoolkit) and it is working very nicely.

The only feedback I have is that instead of doing smoothness (inverse roughness), simply have it as roughness. Unity went against industry standards when they flipped it and it is a pain having to invert mappings from software like substance instead of being able to pack into channels directly for export! Other than that this is a great addition to the repo and we are really glad to see the inclusion of occlusion mapping 👍

Cameron-Micka commented:
@isurfraz That is great to hear! I agree, I'm disappointed Unity went against the grain with smoothness vs roughness.
